File: packages/turbo/ios/RNVisitableViewManager.m
//
//  RNTTurboWebviewManager.m
//  HotwireRNDemo
//
//  Created by Bartłomiej Fryz on 22/06/2022.
//

#pragma clang diagnostic ignored "-Wstrict-prototypes"

#import <React/RCTViewManager.h>
#import <React/RCTEventEmitter.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(RNVisitableViewManager, NSObject)

  RCT_EXPORT_VIEW_PROPERTY(url, NSString)
  RCT_EXPORT_VIEW_PROPERTY(sessionHandle, NSString)
  RCT_EXPORT_VIEW_PROPERTY(applicationNameForUserAgent, NSString)
  RCT_EXPORT_VIEW_PROPERTY(pullToRefreshEnabled, BOOL)
  RCT_EXPORT_VIEW_PROPERTY(scrollEnabled, BOOL)
  RCT_EXPORT_VIEW_PROPERTY(contentInset, NSDictionary)
  RCT_EXPORT_VIEW_PROPERTY(refreshControlTopAnchor, NSNumber)
  RCT_EXPORT_VIEW_PROPERTY(webViewDebuggingEnabled, BOOL)
  RCT_EXPORT_VIEW_PROPERTY(allowsInlineMediaPlayback, BOOL)
  RCT_EXPORT_VIEW_PROPERTY(onVisitProposal, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onOpenExternalUrl, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onMessage, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onLoad, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onError, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onWebAlert, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onWebConfirm, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onFormSubmissionStarted, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onFormSubmissionFinished, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onShowLoading, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onHideLoading, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onContentProcessDidTerminate, RCTDirectEventBlock)

  RCT_EXTERN_METHOD(injectJavaScript: (nonnull NSNumber) node
                    code: (nonnull NSString) code)
  RCT_EXTERN_METHOD(reload: (nonnull NSNumber) node)
  RCT_EXTERN_METHOD(refresh: (nonnull NSNumber) node)
  RCT_EXTERN_METHOD(sendAlertResult: (nonnull NSNumber) node)
  RCT_EXTERN_METHOD(sendConfirmResult: (nonnull NSNumber) node
                    result: (nonnull NSString) code)


@end

File: packages/turbo/ios/RNVisitableViewManager.swift
//
//  RNTTurboWebviewManager.swift
//  HotwireRNDemo
//
//  Created by Bartłomiej Fryz on 24/06/2022.
//

import Foundation
import ReactNativeHotwiredTurboiOS

@objc(RNVisitableViewManager)
class RNVisitableViewManager: RCTViewManager {
  
  override func view() -> UIView! {
    return RNVisitableView()
  }
  
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
    
  @objc 
  func injectJavaScript(_ node: NSNumber, code: NSString) {
    DispatchQueue.main.async {
      let component = self.bridge.uiManager.view(forReactTag: node) as! RNVisitableView
      component.injectJavaScript(code: code)
    }
  }
    
  @objc
  func reload(_ node: NSNumber) {
    DispatchQueue.main.async {
      let component = self.bridge.uiManager.view(forReactTag: node) as! RNVisitableView
      component.reload()
    }
  }

  @objc
  func sendAlertResult(_ node: NSNumber) {
    DispatchQueue.main.async {
      let component = self.bridge.uiManager.view(forReactTag: node) as! RNVisitableView
      component.sendAlertResult()
    }
  }

  @objc
  func sendConfirmResult(_ node: NSNumber, result: NSString) {
    DispatchQueue.main.async {
      let component = self.bridge.uiManager.view(forReactTag: node) as! RNVisitableView
      component.sendConfirmResult(result: result)
    }
  }
    
  @objc
  func clearSnapshotCache(_ node: NSNumber) {
    DispatchQueue.main.async {
      let component = self.bridge.uiManager.view(forReactTag: node) as! RNVisitableView
      component.clearSessionSnapshotCache()
    }
  }
}

